File: src/Reloaded.Memory.Buffers/Internal/BufferAllocator.Linux.cs
using Reloaded.Memory.Buffers.Exceptions;
using Reloaded.Memory.Buffers.Native.Linux;
using Reloaded.Memory.Buffers.Structs.Internal;
using Reloaded.Memory.Buffers.Structs.Params;
using Reloaded.Memory.Buffers.Utilities;
using Reloaded.Memory.Enums;
using Reloaded.Memory.Native.Unix;

namespace Reloaded.Memory.Buffers.Internal;

#pragma warning disable CA1416 // Validate platform compatibility

/// <summary>
/// Windows specific buffer allocator.
/// </summary>
internal static partial class BufferAllocator
{
    // Devirtualized based on target.
    private static LocatorItem AllocateLinux(BufferAllocatorSettings settings)
    {
        for (int x = 0; x < settings.RetryCount; x++)
        {
            // Until we get all of the pages.
            var regions = LinuxMapParser.GetFreeRegions(settings.TargetProcess);
            foreach (MemoryMapEntry region in regions)
            {
                // Exit if we are done iterating.
                if (region.StartAddress > settings.MaxAddress)
                    break;

                // Add the page and increment address iterator to go to next page.
                if (TryAllocateBuffer(region, settings, out var item))
                    return item;
            }
        }

        throw new MemoryBufferAllocationException(settings.MinAddress, settings.MaxAddress, (int)settings.Size);
    }

    private static bool TryAllocateBuffer(MemoryMapEntry entry, BufferAllocatorSettings settings, out LocatorItem result)
    {
        result = default;

        Span<nuint> results = stackalloc nuint[4];
        foreach (var addr in GetPossibleBufferAddresses(settings.MinAddress, settings.MaxAddress, entry.StartAddress, entry.EndAddress, settings.Size, Cached.GetAllocationGranularity(), results))
        {
            // ReSharper disable once RedundantCast
            // MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE = 0x100022
            nint allocated = Posix.mmap(addr, (nuint)settings.Size, (int)MemoryProtection.ReadWriteExecute, 0x100022, -1, 0);

            if (allocated == -1)
                continue;

            // Just in case, for older kernels;
            if ((nuint)allocated != addr)
            {
                Posix.munmap((nuint)allocated, settings.Size);
                continue;
            }

            result = new LocatorItem((nuint)allocated, settings.Size);
            return true;
        }

        return false;
    }
}

File: src-rust/src/internal/buffer_allocator_linux.rs
use crate::internal::buffer_allocator::get_possible_buffer_addresses;
use crate::structs::errors::BufferAllocationError;
use crate::structs::internal::LocatorItem;
use crate::structs::params::BufferAllocatorSettings;
use crate::utilities::cached::CACHED;
use crate::utilities::linux_map_parser::{get_free_regions_from_process_id, MemoryMapEntry};
use libc::{
    mmap, munmap, MAP_ANONYMOUS, MAP_FIXED_NOREPLACE, MAP_PRIVATE, PROT_EXEC, PROT_READ, PROT_WRITE,
};

// Implementation //
pub fn allocate_linux(
    settings: &BufferAllocatorSettings,
) -> Result<LocatorItem, BufferAllocationError> {
    for _ in 0..settings.retry_count {
        let regions = get_free_regions_from_process_id(settings.target_process_id as i32);
        for region in regions {
            if region.start_address > settings.max_address {
                break;
            }

            unsafe {
                match try_allocate_buffer(&region, &settings) {
                    Ok(item) => return Ok(item),
                    Err(_) => continue,
                }
            }
        }
    }

    Err(BufferAllocationError::new(
        *settings,
        "Failed to allocate buffer on Linux",
    ))
}

unsafe fn try_allocate_buffer(
    entry: &MemoryMapEntry,
    settings: &BufferAllocatorSettings,
) -> Result<LocatorItem, &'static str> {
    let buffer: &mut [usize; 4] = &mut [0; 4];

    for addr in get_possible_buffer_addresses(
        settings.min_address,
        settings.max_address,
        entry.start_address,
        entry.end_address,
        settings.size as usize,
        CACHED.get_allocation_granularity() as usize,
        buffer,
    ) {
        let allocated = mmap(
            *addr as *mut _,
            settings.size as usize,
            PROT_READ | PROT_WRITE | PROT_EXEC,
            MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED_NOREPLACE,
            -1,
            0,
        );

        if allocated.is_null() {
            continue;
        }

        if allocated as usize != *addr {
            munmap(allocated, settings.size as usize);
            continue;
        }

        return Ok(LocatorItem::new(allocated as usize, settings.size));
    }

    Err("Failed to allocate buffer")
}
